### DISCUSSION TITLE
How do you generate a "yield return null;" for an IEnumerator method?

### DISCUSSION THREAD

**Discussion by rickomax** [jbevain/cecil#897] [Q&A] [answered]
Hi!

I'm trying to use Mono.Cecil to duplicate some methods based on custom Attributes to make these copies use IEnumerator (so I can use them as Unity Coroutines).

I can load the original libraries with Cecil, find the methods, duplicate them, etc.

The problem is:

I have to replace some placeholder method calls with a yield return null;, which seems to require generating more IL code than I expected.

Does anyone know how that could be represented as IL code with Mono.Cecil?

I've been using https://sharplab.io/ to understand the pattern, but it seems like it could be more complex.

**Comment by rickomax** [comment]
So, something like:
```csharp
[CoroutineMethod]
private int TestMethod(AssetLoaderContext assetLoaderContext, int a, int b) {
    if (assetLoaderContext.CheckMainThreadStall()) {

    }
    return a + b;
}
```
Could be duplicated to:
```csharp
private IEnumerator TestMethod_Coroutine(Stack result, AssetLoaderContext assetLoaderContext, int a, int b) {
    if (assetLoaderContext.CheckMainThreadStall()) {
        yield return null;
    }
    result.Push(a + b);
}
```

**Comment by ltrzesniewski** [comment, marked as answer]
There is no direct IL equivalent to `yield return`. Methods that use `yield` are rewritten by the compiler as finite state machines, with state changes on `yield` statements.

I'm afraid this transformation is not exactly straightforward (*especially* starting from the low-level IL representation), but you can take a look at the [Roslyn implementation](https://github.com/dotnet/roslyn/tree/main/src/Compilers/CSharp/Portable/Lowering/IteratorRewriter) if you'd like to replicate it. Comments there can be insightful, like [this one](https://github.com/dotnet/roslyn/blob/af08c178186130765ad21174bad97b7899b491a5/src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorMethodToStateMachineRewriter.cs#L322-L329):

```csharp
//     yield return expression;
// is translated to
//     this.current = expression;
//     this.state = <next_state>;
//     return true;
//     <next_state_label>: ;
//     <hidden sequence point>
//     this.state = finalizeState;
```

**Reply by rickomax** [reply]
I appreciate it!
I ended up using Roslyn Code Generator for the task. It does not provide me strong type references (or at least, I still need help finding how to get them from parsed code), but it makes such compiler-specific conversion more trivial.
